src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -436,18 +436,35 @@ void GcsActorManager::HandleRestartActor(rpc::RestartActorRequest request,
   };
   auto pending_restart_iter = actor_to_restart_callbacks_.find(actor_id);
   if (pending_restart_iter != actor_to_restart_callbacks_.end()) {
+    // This happens when the RestartActor rpc is received and being handled
+    // and then the connection is lost and the caller resends the same request.
     pending_restart_iter->second.emplace_back(std::move(success_callback));
     return;
   }
 
   auto actor = iter->second;
-  if (request.num_restarts() <= actor->GetActorTableData().num_restarts()) {
-    // This is a stale message.
+  if (request.num_restarts_due_to_lineage_reconstruction() <=
+      actor->GetActorTableData().num_restarts_due_to_lineage_reconstruction()) {
```

Comment:
Actual fix: idempotency token should be `num_restarts_due_to_lineage_reconstruction` instead of `num_restarts`